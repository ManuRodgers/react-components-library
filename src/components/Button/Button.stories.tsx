import { action } from '@storybook/addon-actions';
import React from 'react';

import Button from './Button';

export default {
  title: 'Button Component',
  component: Button,
};

export const DefaultButton = (): JSX.Element => (
  <Button onClick={action('clicked')}>Default Button</Button>
);
