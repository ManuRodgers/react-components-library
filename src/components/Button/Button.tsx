import classnames from 'classnames';
import React, { AnchorHTMLAttributes, ButtonHTMLAttributes, FC } from 'react';

export type ButtonSize = 'lg' | 'sm';
export type ButtonType = 'primary' | 'default' | 'danger' | 'link';

interface BaseButtonProps {
  className?: string;
  /**set Button  disabled*/
  disabled: boolean;
  /**set Button size */
  size?: ButtonSize;
  /**set Button type */
  btnType: ButtonType;
  children: React.ReactNode;
  href?: string;
}

type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLElement>;
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>;
export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>;

const Button: FC<ButtonProps> = ({
  children,
  btnType,
  className,
  disabled,
  size,
  href,
  ...restProps
}: ButtonProps): JSX.Element => {
  const classNames = classnames('btn', className, {
    [`btn-${btnType}`]: btnType,
    [`btn-${size}`]: size,
    disabled: btnType === 'link' && disabled,
  });
  if (btnType === 'link' && href) {
    return (
      <a href={href} className={classNames} {...restProps}>
        {children}
      </a>
    );
  } else {
    return (
      <button disabled={disabled} className={classNames} {...restProps}>
        {children}
      </button>
    );
  }
};
Button.defaultProps = {
  btnType: 'default',
  disabled: false,
};
export default Button;
