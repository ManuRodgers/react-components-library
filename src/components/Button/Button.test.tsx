import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

import Button, { ButtonProps } from './Button';

const defaultProps = {
  onClick: jest.fn(),
};
const testProps: ButtonProps = {
  btnType: 'primary',
  size: 'lg',
  className: 'manu',
};
const disabledProps: ButtonProps = {
  disabled: true,
  onClick: jest.fn(),
};
describe('Button Component', () => {
  afterEach(cleanup);

  it('should render the correct default button', () => {
    // assignment
    const wrapper = render(<Button {...defaultProps}>Nice</Button>);
    const element = wrapper.getByText('Nice') as HTMLButtonElement;

    //  assertion
    expect(element).toBeInTheDocument();
    expect(element.tagName).toEqual('BUTTON');
    expect(element).toHaveClass('btn btn-default');
    expect(element.disabled).toBeFalsy();
    // action
    fireEvent.click(element);
    expect(defaultProps.onClick).toHaveBeenCalled();
  });
  it('should render the correct component based on different props', () => {
    const wrapper = render(<Button {...testProps}>Nice</Button>);
    const element = wrapper.getByText('Nice');
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('btn-primary btn-lg manu');
  });
  it('should render a link when btnType equals link and href is provided', () => {
    const wrapper = render(
      <Button btnType="link" href="http://dummyurl">
        Link
      </Button>
    );
    const element = wrapper.getByText('Link');
    expect(element).toBeInTheDocument();
    expect(element.tagName).toEqual('A');
    expect(element).toHaveClass('btn btn-link');
  });
  it('should render disabled button when disabled set to true', () => {
    const wrapper = render(<Button {...disabledProps}>Nice</Button>);
    const element = wrapper.getByText('Nice') as HTMLButtonElement;
    expect(element).toBeInTheDocument();
    expect(element.disabled).toBeTruthy();
    fireEvent.click(element);
    expect(disabledProps.onClick).not.toHaveBeenCalled();
  });
  it('matches snapshot', async () => {
    const { asFragment } = render(<Button {...defaultProps}>Nice</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});
