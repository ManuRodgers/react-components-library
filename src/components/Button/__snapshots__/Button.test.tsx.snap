// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-default"
  >
    Nice
  </button>
</DocumentFragment>
`;
